File: 2023/src/nim/dayOne.nim
# Copyright (C) 2023 Rémy Cases
# See LICENSE file for extended copyright information.
# This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

import os, algorithm

const digits = { 1 .. 9 }
const zero = int('0')
const validAnswer = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

proc validAnswerToInt(va: string, digit: var int) =
    if va == "1" or va == "one":
        digit = 1
    elif va == "2" or va == "two":
        digit = 2
    elif va == "3" or va == "three":
        digit = 3
    elif va == "4" or va == "four":
        digit = 4
    elif va == "5" or va == "five":
        digit = 5
    elif va == "6" or va == "six":
        digit = 6
    elif va == "7" or va == "seven":
        digit = 7
    elif va == "8" or va == "eight":
        digit = 8
    elif va == "9" or va == "nine":
        digit = 9

proc parseNaive(line: string, firstDigit: var int, secondDigit: var int) =
    if line.len == 0: return
    for c in line:
        if int(c)-zero in digits:
            firstDigit = int(c)-zero
            break
    
    for c in line.reversed():
        if int(c)-zero in digits:
            secondDigit = int(c)-zero
            break

proc parseWithString(line: string, firstDigit: var int, secondDigit: var int) =
    var low = line.low
    var high = low
    let len = line.len
    while(low < len):
        if line[low .. high] in validAnswer:
            validAnswerToInt(line[low .. high], firstDigit)
            break
        high = high + 1
        if high == len or high - low > 5: # there is no element bigger than a six letters string
            low = low + 1
            high = low
    
    low = line.high
    high = low
    while(high >= 0):
        if line[low .. high] in validAnswer:
            validAnswerToInt(line[low .. high], secondDigit)
            break
        low = low - 1
        if low == -1 or high - low > 5:
            high = high - 1
            low = high


proc readFileAndCalibration(filename: string, option = '1') =
    var firstDigit: int = 0
    var secondDigit: int = 0
    var calibrationValue: int = 0

    let parse = case option:
    of '1':
        parseNaive
    of '2':
        parseWithString
    else:
        return

    for line in filename.lines:
        parse(line, firstDigit, secondDigit)
        calibrationValue += firstDigit*10 + secondDigit

    echo "CALIBRATION VALUE: ", calibrationValue

proc main*(part: char) =
    const file = "2023/data/input_day_one"
    let filename = getCurrentDir() / file
    readFileAndCalibration(filename, part)

when isMainModule:
    main('1')

File: 2023/src/rust/day_one.rs
// Copyright (C) 2023 Rémy Cases
// See LICENSE file for extended copyright information.
// This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

use std::path::Path;
use std::io::{Error, ErrorKind};
use std::env;
use aoc_utils::*;

const VALID_STRING: [&str; 18] = ["1", "2", "3", "4", "5", "6", "7", "8", "9", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

type FnParse = fn(io::Result<String>, &mut i32, &mut i32) -> io::Result<()>;

fn convert_text_int(text: &str, digit: &mut i32) -> io::Result<()> {
    if text == "1" || text == "one" { *digit = 1; Ok(()) }
    else if text == "2" || text == "two" { *digit = 2; Ok(()) }
    else if text == "3" || text == "three" { *digit = 3; Ok(()) }
    else if text == "4" || text == "four" { *digit = 4; Ok(()) }
    else if text == "5" || text == "five" { *digit = 5; Ok(()) }
    else if text == "6" || text == "six" { *digit = 6; Ok(()) }
    else if text == "7" || text == "seven" { *digit = 7; Ok(()) }
    else if text == "8" || text == "eight" { *digit = 8; Ok(()) }
    else if text == "9" || text == "nine" { *digit = 9; Ok(()) }
    else { Err(Error::new(ErrorKind::InvalidInput, "Invalid string")) }
}

fn parse_digits(line: io::Result<String>, first_digit: &mut i32, second_digit: &mut i32) -> io::Result<()> {
    let binding_line = line?;
    let mut parsed_line = binding_line.chars().filter_map(|l| l.to_digit(10));
    let mut parsed_reversed_line = binding_line.chars().rev().filter_map(|l| l.to_digit(10));

    *first_digit = parsed_line.next().ok_or(Error::new(ErrorKind::NotFound, "Digit not found"))? as i32;
    *second_digit = parsed_reversed_line.next().ok_or(Error::new(ErrorKind::NotFound, "Digit not found"))? as i32;
    
    Ok(())
}

fn parse_digits_and_name(line: io::Result<String>, first_digit: &mut i32, second_digit: &mut i32) -> io::Result<()> {
    let binding_line = line?;
    let it_line_for_min = VALID_STRING.iter().filter_map(|s| binding_line.find(s).map(|r| (r, s.len())));
    let it_line_for_max = VALID_STRING.iter().filter_map(|s| binding_line.rfind(s).map(|r| (r, s.len())));

    let min_index = it_line_for_min.min_by(|(x, _), (z, _)| x.cmp(z)).ok_or(Error::new(ErrorKind::NotFound, "Digit not found"))?;
    let max_index = it_line_for_max.max_by(|(x, _), (z, _)| x.cmp(z)).ok_or(Error::new(ErrorKind::NotFound, "Digit not found"))?;
 
    convert_text_int(&binding_line[min_index.0..min_index.0 + min_index.1], first_digit)?;
    convert_text_int(&binding_line[max_index.0..max_index.0 + max_index.1], second_digit)?;

    Ok(())
}

fn read_file_and_compute_calibration(file_path: &Path, part: argparse::ArgPart) -> io::Result<()> {
    let lines = io::line_iterator(file_path)?;
    let mut first_digit_calibration = 0;
    let mut second_digit_calibration = 0;
    let mut sum_calibration = 0;

    let tryparse: io::Result<FnParse> = match part {
        argparse::ArgPart::PartOne => Ok(parse_digits),
        argparse::ArgPart::PartTwo => Ok(parse_digits_and_name),
    };

    let parse = tryparse?;
    for line in lines { 
        parse(line, &mut first_digit_calibration, &mut second_digit_calibration)?;
        sum_calibration += first_digit_calibration * 10 + second_digit_calibration;
    }
  
    println!("CALIBRATION VALUE: {:}", sum_calibration);
    Ok(())
}

pub fn main(part: argparse::ArgPart) -> io::Result<()> {
    let filename = env::current_dir()?.join("2023").join("data").join("input_day_one");
    read_file_and_compute_calibration(&filename, part)?;
    Ok(())
}
